modules/guides/pages/delivery_semantics.adoc

```diff
@@ -0,0 +1,43 @@
+= Message Delivery Semantics
+:description: Learn about Redpanda Connect's transactional model and error handling
+
+Redpanda Connect (RPCN) guarantees strong message delivery semantics across its data pipelines by leveraging a transactional model centered around message batch acknowledgment. This ensures reliable, scalable, and predictable data movement between input and output systems, with particular support for *at-least-once delivery* — without requiring intermediate message persistence.
+
+This topic describes how Redpanda Connect manages message delivery across the Input → Pipeline → Output architecture.
+
+image::../../shared/images/delivery-semantics-arch.svg[]
+
+== Delivery Guarantees
+
+Associated with every batch of messages is an `ack()` function. It's a callback function provided by the input and its the outputs job to call the `ack()` function to inform the input of the success ("ack") or failure (negative ack or "nack") of the batch. The input can then decide what to do, e.g. auto-retry nacks, propagate the nack upstream to input connectors that support it (Google PubSub), or drop the batch. When connecting to *at-least-once* inputs and outputs, Redpanda Connect is able to guarantee *at-least-once* delivery without needing to persist messages during transit.
+
+Key Benefits:
+
+* Messages will not be lost unless explicitly dropped in configuration
+* Ensures every message is processed at-least-once, with optional retry logic for failures
+* Duplicates may occur in failure recovery scenarios — typical of at-least-once systems
+* Delivery status is communicated back to the input, enabling robust retry and propagation strategies
+* No need for disk-based check-pointing — reducing latency and infrastructure complexity
+
+== Caveats to At-Least-Once
+
+At-least-once delivery in a Redpanda Connect pipeline means every message will be delivered at-least-once even if failures occur (which might result in some duplicates, but no data is lost). Introducing a distributed or externally persisted state into a pipeline can compromise this guarantee. The pipeline's built-in fault tolerance assumes that if an output fails, the message will be retried or reprocessed upon recovery. However, an external state (such as a remote cache or database used within the pipeline) may not reset on failure. It could retain an indication that a message was processed even if that message was never successfully processed by the output, and this mismatch can lead to message loss during recovery.
```

Comment:
```suggestion
At-least-once delivery in a Redpanda Connect pipeline means every message is delivered at-least-once even if failures occur (which could result in some duplicates, but no lost data). Introducing a distributed or externally persisted state into a pipeline can compromise this guarantee. The pipeline's built-in fault tolerance is based on the assumption that if an output fails, the message will be retried or reprocessed upon recovery. However, an external state (such as a remote cache or database used within the pipeline) may not reset on failure. It could retain an indication that a message was processed even if that message was never successfully processed by the output, and this mismatch can lead to message loss during recovery.
```